consensus/src/alerts/handler.rs

```diff
@@ -0,0 +1,752 @@
+/// https://cardinal-cryptography.github.io/AlephBFT/how_alephbft_does_it.html Section 2.5 and
+/// https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html and to the Aleph
+/// paper https://arxiv.org/abs/1908.05156 Appendix A1 for a discussion.
+pub struct Handler<H: Hasher, D: Data, MK: MultiKeychain> {
+    session_id: SessionId,
+    keychain: MK,
+    known_forkers: HashMap<NodeIndex, ForkProof<H, D, MK::Signature>>,
+    known_alerts: KnownAlerts<H, D, MK>,
+    known_rmcs: HashMap<(NodeIndex, NodeIndex), H::Hash>,
+    pub exiting: bool,
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> Handler<H, D, MK> {
+    pub fn new(keychain: MK, config: AlertConfig) -> Self {
+        Self {
+            session_id: config.session_id,
+            keychain,
+            known_forkers: HashMap::new(),
+            known_alerts: HashMap::new(),
+            known_rmcs: HashMap::new(),
+            exiting: false,
+        }
+    }
+
+    pub fn index(&self) -> NodeIndex {
+        self.keychain.index()
+    }
+
+    pub fn is_forker(&self, forker: NodeIndex) -> bool {
+        self.known_forkers.contains_key(&forker)
+    }
+
+    pub fn on_new_forker_detected(
+        &mut self,
+        forker: NodeIndex,
+        proof: ForkProof<H, D, MK::Signature>,
+    ) {
+        self.known_forkers.insert(forker, proof);
+    }
+
+    // Correctness rules:
+    // 1) All units must be created by forker
+    // 2) All units must come from different rounds
+    // 3) There must be fewer of them than the maximum defined in the configuration.
+    // Note that these units will have to be validated before being used in the consensus.
+    // This is alright, if someone uses their alert to commit to incorrect units it's their own
+    // problem.
+    pub fn correct_commitment(
+        &self,
+        forker: NodeIndex,
+        units: &[UncheckedSignedUnit<H, D, MK::Signature>],
+    ) -> bool {
+        let mut rounds = HashSet::new();
+        for u in units {
+            let u = match u.clone().check(&self.keychain) {
+                Ok(u) => u,
+                Err(_) => {
+                    warn!(target: "AlephBFT-alerter", "{:?} One of the units is incorrectly signed.", self.index());
+                    return false;
+                }
+            };
+            let full_unit = u.as_signable();
+            if full_unit.creator() != forker {
+                warn!(target: "AlephBFT-alerter", "{:?} One of the units {:?} has wrong creator.", self.index(), full_unit);
+                return false;
+            }
+            if rounds.contains(&full_unit.round()) {
+                warn!(target: "AlephBFT-alerter", "{:?} Two or more alerted units have the same round {:?}.", self.index(), full_unit.round());
+                return false;
+            }
+            rounds.insert(full_unit.round());
+        }
+        true
+    }
+
+    pub fn who_is_forking(&self, proof: &ForkProof<H, D, MK::Signature>) -> Option<NodeIndex> {
```

Comment:
Should this even be `pub`? I'm pretty sure not all the methods here should, please un`pub` the ones that are not used externally.